src/sinks/datadog/metrics/encoder.rs

```diff
@@ -502,6 +528,101 @@ fn encode_timestamp(timestamp: Option<DateTime<Utc>>) -> i64 {
+// Some sources such as `kafka`, `nats`, `redis` for example, are only capable of receiving metrics
+// with the `native` or `native_json` codec. In such cases we intentionally do not set the origin
+// metadata here, because the true origin will have already been determined to be a pass-through.
+fn source_type_to_service(source_type: &str) -> Option<u32> {
+    if source_type == "apache_metrics" {
+        Some(17)
+    } else if source_type == "aws_ecs_metrics" {
+        Some(209)
+    } else if source_type == "eventstoredb_metrics" {
+        Some(210)
+    } else if source_type == "host_metrics" {
+        Some(211)
+    } else if source_type == "internal_metrics" {
+        Some(212)
+    } else if source_type == "mongodb_metrics" {
+        Some(111)
+    } else if source_type == "nginx_metrics" {
+        Some(117)
+    } else if source_type == "open_telemetry" {
+        Some(213)
+    } else if source_type == "postgresql_metrics" {
+        Some(128)
+    } else if source_type == "prometheus_remote_write" {
+        Some(214)
+    } else if source_type == "prometheus_scrape" {
+        Some(215)
+    } else if source_type == "statsd" {
+        Some(153)
+    } else {
+        None
+    }
+}
+
+/// Determine the correct Origin metadata values to use depending on if they have been
+/// set already upstream or not. The generalized struct `DatadogMetricOriginMetadata` is
+/// utilized in this function, which allows the series and sketch encoding to call and map
+/// the result appropriately for the given protocol they operate on.
+fn generate_origin_metadata(
+    maybe_pass_through: Option<&DatadogMetricOriginMetadata>,
+    maybe_source_type: Option<&Arc<String>>,
+) -> Option<DatadogMetricOriginMetadata> {
+    let origin_product_value = option_env!("ORIGIN_PRODUCT")
+        .map(|p| {
+            p.parse::<u32>()
+                .expect("Env var ORIGIN_PRODUCT must be an unsigned 32 bit integer.")
+        })
+        .unwrap_or(14);
+
+    let origin_category_value = 11;
+
+    let no_value = 0;
+
+    // An upstream vector source or a transform has set the origin metadata already.
+    // Currently this is only possible by these scenarios:
+    //     - `datadog_agent` source receiving the metadata on ingested metrics
+    //     - `vector` source receiving events with EventMetadata that already has the origins set
+    //     - `log_to_metric` transform set the OriginService in the EventMetadata when it creates
+    //        the new metric.
+    if let Some(pass_through) = maybe_pass_through {
+        Some(
+            DatadogMetricOriginMetadata::default()
+                .with_product(pass_through.product().unwrap_or(origin_product_value))
+                .with_category(pass_through.category().unwrap_or(origin_category_value))
+                .with_service(pass_through.service().unwrap_or(no_value)),
+        )
+
+    // no metadata has been set upstream
+    } else {
+        maybe_source_type.and_then(|source_type| {
```

Comment:
Is my understanding correct that if the Datadog Agent source isn't sending these headers we will also avoid sending them via `maybe_source_type` being empty here?

Reply:
Ah, gotcha, I missed that. It does seem like readers would be more likely to associate it with that arm of the if statement if it were closer (or an additional comment was included there).

Reply:
In reference to @StephenWakely's comment below, I think we should set the origin metadata if the `source_type` is not the datadog agent source but unknown. In that case, we would set the service to unknown as well. This would be slightly more robust than not setting the metadata at all in those scenarios in my opinion..